solutionbox/structured_data/datalab_solutions/structured_data/predict/predict.py

```diff
@@ -0,0 +1,434 @@
+  """A DoFn to fix missing target columns."""
+
+  def __init__(self, trained_model_dir):
+    """Reads the schema file and extracted the expected number of columns.
+
+    Args:
+      trained_model_dir: path to model.
+
+    Raises:
+      ValueError: if schema.json not found in trained_model_dir
+    """
+    from tensorflow.python.lib.io import file_io
+    import json
+    import os
+
+    schema_path = os.path.join(trained_model_dir, 'schema.json')
+    if not file_io.file_exists(schema_path):
+      raise ValueError('schema.json missing from %s' % schema_path)
+    schema = json.loads(file_io.read_file_to_string(schema_path))
+    self._num_expected_columns = len(schema)
+
+  def process(self, context):
+    import logging
+    import apache_beam as beam
+
+    num_columns = len(context.element.split(','))
+    if num_columns == self._num_expected_columns:
+      yield context.element
+    elif num_columns + 1 == self._num_expected_columns:
+      yield ',' + context.element
+    else:
+      logging.error('Got an unexpected number of columns from [%s].' %
+                    context.element)
+      yield beam.pvalue.SideOutputValue('errors',
+                                        ('bad columns', context.element))
+
+
+class EmitAsBatchDoFn(beam.DoFn):
+  """A DoFn that buffers the records and emits them batch by batch."""
+
+  def __init__(self, batch_size):
+    """Constructor of EmitAsBatchDoFn beam.DoFn class.
+
+    Args:
+      batch_size: the max size we want to buffer the records before emitting.
+    """
+    self._batch_size = batch_size
+    self._cached = []
+
+  def process(self, context):
+    self._cached.append(context.element)
+    if len(self._cached) >= self._batch_size:
+      emit = self._cached
+      self._cached = []
+      yield emit
+
+  def finish_bundle(self, context):
+    if len(self._cached) > 0:  # pylint: disable=g-explicit-length-test
+      yield self._cached
+
+
+class RunGraphDoFn(beam.DoFn):
+  """A DoFn for running the TF graph."""
+
+  def __init__(self, trained_model_dir):
+    self._trained_model_dir = trained_model_dir
+    self._session = None
+
+  def _setup_prediction(self):
+    from tensorflow.contrib.session_bundle import session_bundle
+    import json
+
+    self._session, _ = session_bundle.load_session_bundle_from_path(
+        self._trained_model_dir)
+
+    # input_alias_map {'input_csv_string': tensor_name}
+    self._input_alias_map = json.loads(
+        self._session.graph.get_collection('inputs')[0])
+
+    # output_alias_map {'target_from_input': tensor_name, 'key': ...}
+    self._output_alias_map = json.loads(
+        self._session.graph.get_collection('outputs')[0])
+
+    self._aliases, self._tensor_names = zip(*self._output_alias_map.items())
+
+  def process(self, context):
+    import collections
+    import logging
+    import apache_beam as beam   
+
+    num_in_batch = 0
+    try:
+      if not self._session:
+        self._setup_prediction()
```

Comment:
Maybe do that in start_bundle? Also need to close the session in finish_bundle, otherwise there will be a leak in local run (Once I hit real issues by not closing previous session).